build-tools/build.gradle.kts

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+import org.jetbrains.kotlin.gradle.plugin.KotlinSourceSet
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import java.util.Properties
+
+plugins {
+    // We explicitly configure versions of plugins in settings.gradle.kts.
+    // due to https://github.com/gradle/gradle/issues/1697.
+    id("kotlin")
+    id("kotlinx-serialization")
+    groovy
+    `java-gradle-plugin`
+}
+
+val rootProperties = Properties().apply {
+    rootDir.resolve("../gradle.properties").reader().use(::load)
+}
+
+val kotlinVersion: String by rootProperties
+val kotlinCompilerRepo: String by rootProperties
+val buildKotlinVersion: String by rootProperties
+val buildKotlinCompilerRepo: String by rootProperties
+val konanVersion: String by rootProperties
+
+group = "org.jetbrains.kotlin"
+version = konanVersion
+
+repositories {
+    maven(kotlinCompilerRepo)
+    maven(buildKotlinCompilerRepo)
+    maven("https://cache-redirector.jetbrains.com/maven-central")
+    maven("https://kotlin.bintray.com/kotlinx")
+}
+
+dependencies {
+    compileOnly(gradleApi())
+
+    implementation("org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlinVersion")
+    implementation("org.jetbrains.kotlin:kotlin-stdlib:$kotlinVersion")
+    implementation("org.jetbrains.kotlin:kotlin-reflect:$kotlinVersion")
+    implementation("com.ullink.slack:simpleslackapi:1.2.0")
+    implementation("org.jetbrains.kotlinx:kotlinx-serialization-runtime:0.10.0")
+
+    implementation("io.ktor:ktor-client-auth:1.2.1")
+    implementation("io.ktor:ktor-client-core:1.2.1")
+    implementation("io.ktor:ktor-client-cio:1.2.1")
+
+    api("org.jetbrains.kotlin:kotlin-native-utils:$kotlinVersion")
+
+    // Located in <repo root>/shared and always provided by the composite build.
+    api("org.jetbrains.kotlin:kotlin-native-shared:$konanVersion")
+}
+
+sourceSets["main"].withConvention(KotlinSourceSet::class) {
```

Comment:
Why this sourceSet in't defined like this? 
```
sourceSets { kotlin {  srcDir () } } 
```
What does `withConvention` do?

Reply:
The `kotlin` source directory set is added as a dynamic property using the corresponding Gradle machinery. But there is no extension getter allowing us just write something like this

```
sourceSets["main"].kotlin // kotlin property isn't available
```

`withConvention` is a way to access this property [recommended by the documentation](https://kotlinlang.org/docs/reference/using-gradle.html#targeting-the-jvm).